### PR TITLE
Support `(type) satisfies never;`

### PR DESCRIPTION
**Pull request by seiyab**
## Description

Fixes https://github.com/prettier/prettier/issues/15491.

## Checklist

- [x] I’ve added tests to confirm my change works.
- [x] ~(If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory).~
- [x] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/main/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

**Comment by fisker** [comment]
**Prettier pr-15514**
[Playground link](https://deploy-preview-15514--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEAKGBPADnAlAAgEMBnfAMwgn2MJgEtiy65So4A3OAJwG58QANCAhZ60YslCEuXCAHcACtIQSUhADZzCGCUIBGXQmADWcGAGVCAWzgAZOm2RkNxOPsMmz5rEYcBzZBguAFc3EFcrOkCQsOJ-dTgARWCIeCcXMIArYgAPc3iklLSkZ3VXIQBHIrgFWSxVEBIAWjY4ABN2wRAgwjp1fwBhCCsrQmRG9XUuuKg-BIBBGCC6PWD4BW57RxKMoQALGCt1AHU9unhiHzA4cxVzunZzjHGwYl0QdlCASSgO2HMwFw6KJ5r9zJgEukymEsLJXMdDFhxrCWNxOF0HK4uDBaoQ-KMoeUQD4uFjxpgcMRAcCYF1YQ4YMc6G0YHtkAAOAAMQi4cCqdF5uPxYx20KEMEIeiZLLZSAATEJgq4ACqS1SlIlwKx6dodNq2QizYJ4uAAMQgXFGS3840IawgIAAvo6gA)

```sh
--parser typescript
```

**Input:**

```tsx
(type) as foo satisfies never; 
```

**Output:**

```tsx
type as foo satisfies never;

```

**Second Output:**

```tsx
SyntaxError: '=' expected. (1:9)
> 1 | type as foo satisfies never;
    |         ^
  2 |
```

**Comment by fisker** [comment]
Flow has another AST not checked.

**Prettier pr-15514**
[Playground link](https://deploy-preview-15514--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEAKGBPADnAlAAgEMBnfSKYmEAGhAixgEtpjlRCAnDiAdwAVOCVikIAbHoQytaAIw6EwAazgwAyoQC2cADKMocZADMxxOLPlKVqrAr0BzZDA4BXMyFMbGjl2+L3RcACKzhDwRiZuAFbEAB6q-kEhYUjGoqa0AI5JcHzcWMIgJAC0+nAAJuU0IE6EjKL2AMIQGhqEyIWiolV+UHYBAIIwTowyzvB8cBy6+uFpbgAWMBqiAOrzjPDENmBwqkIbjABuGxjtYMTSIIeuAJJQFbCqYByMDP33qpgBs+kgWNymFbyLDtf5wUwcQ4GWh6CEwXKEOytH5uGwcCHtQyiXhVf56GArRhlGDzZAADgADLQOHAsowaQikW0UhFaDBCDJCcTSUgAEy0ZymAAqHOEqV+cA0MnKFTK2kIvWciLgADEIBxWkN7O1CGMICAAL4GoA)

```sh
--parser flow
```

**Input:**

```jsx
(type) as const
```

**Output:**

```jsx
type as const;

```

**Second Output:**

```jsx
SyntaxError: Unexpected identifier, expected the token `=` (1:9)
> 1 | type as const;
    |         ^^^^^
  2 |
```

**Comment by seiyab** [comment]
@fisker Thank you for pointing out problems! Please take another look.

File: src/language-js/needs-parens.js [hunk 1]
```diff
@@ -110,6 +110,37 @@ function needsParens(path, options) {
...
+          case "TSSatisfiesExpression":
+          case "SatisfiesExpression":
+          case "TSAsExpression":
+          case "AsExpression":
+          case "AsConstExpression":
+            return false;
+          default:
+            return true;
+        }
+      });
```
Comment (fisker) [inline src/language-js/needs-parens.js:126]: There is an `isBinaryCastExpression ` utility we can use.

File: src/language-js/needs-parens.js [hunk 2]
```diff
@@ -110,6 +110,37 @@ function needsParens(path, options) {
...
+      if (
+        ancestorNeitherAsNorSatisfies !== parent &&
+        ancestorNeitherAsNorSatisfies.type === "ExpressionStatement"
+      ) {
+        switch (node.name) {
```
Comment (fisker) [inline src/language-js/needs-parens.js:131]: We should move the name check to line 114, so we don't need search for ancestors when they are not keywords.

**Review by fisker — APPROVED** [review]
Good job.

**Review by sosukesuzuki — APPROVED** [review]

**Comment by seiyab** [comment]
Thank you for your review 🚀 .
I would be grateful if you could also review https://github.com/prettier/prettier/pull/15171.

**Pull request merged**